autopilot.go

```diff
@@ -0,0 +1,211 @@
+	// state should be made but instead a new state is created and set to this field
+	// while holding the stateLock.
+	state *State
+	// stateLock is meant to only protect the state field. This just prevents
+	// the periodic state update and consumers requesting the autopilot state from
+	// racing.
+	stateLock sync.RWMutex
+
+	// startTime is recorded so that we can make better determinations about server
+	// stability during the initial period of time after autopilot first starts.
+	// If autopilot has just started the default behavior to check if a server is
+	// stable will not work as it will ensure the server has been healthy for
+	// the configured server stabilization time. If that configure time is longer
+	// than the amount of time autopilot has been running you can run into issues
+	// with leadership flapping during some scenarios where a cluster is being
+	// brought up.
+	startTime time.Time
+
+	// running is a simple bool to indicate whether the go routines to actually
+	// execute autopilot are currently running
+	running bool
+
+	// removeDeadCh is used to trigger the running autopilot go routines to
+	// find and remove any dead/failed servers
+	removeDeadCh chan struct{}
+
+	// reconcileCh is used to trigger an immediate round of reconciliation.
+	reconcileCh chan struct{}
+
+	// shutdown is a function that can be execute to shutdown a running
+	// autopilot's go routines.
+	shutdown context.CancelFunc
+	// done is a chan that will be closed when the running autopilot go
+	// routines have exited. Technically closing it is the very last
+	// thing done in the go routine but at that point enough state has
+	// been cleaned up that we would then allow it to be started
+	// immediately afterward
+	done chan struct{}
+
+	// runLock is meant to protect all of the fields regarding coordination
+	// of whether the autopilot go routines are running and
+	// starting/stopping them.
+	runLock sync.Mutex
+}
+
+// New will create a new Autopilot instance utilizing the given Raft and Delegate.
+// If the WithPromoter option is not provided the default StablePromoter will
+// be used.
+func New(raft Raft, delegate ApplicationIntegration, options ...Option) *Autopilot {
+	a := &Autopilot{
+		raft:     raft,
+		delegate: delegate,
+		promoter: DefaultPromoter(),
+		logger:   hclog.Default().Named("autopilot"),
+		// should this be buffered?
+		removeDeadCh:      make(chan struct{}, 1),
+		reconcileInterval: DefaultReconcileInterval,
+		updateInterval:    DefaultUpdateInterval,
+		time:              &runtimeTimeProvider{},
+	}
+
+	for _, opt := range options {
+		opt(a)
+	}
+
+	return a
+}
+
+// RemoveDeadServers will trigger and immediate removal of dead/failed servers.
+func (a *Autopilot) RemoveDeadServers() {
+	select {
+	case a.removeDeadCh <- struct{}{}:
+	default:
+	}
+}
+
+// GetState retrieves the current autopilot State
+func (a *Autopilot) GetState() *State {
+	a.stateLock.Lock()
+	defer a.stateLock.Unlock()
+	return a.state
```

Comment:
This is returning a pointer to `State`, which means it can be altered afterwards. I don't think the lock doesn't cover all the cases, like concurrently writing at the same time. Or am I missing something?

Reply:
This similar to data in memdb should be considered read-only. If Go could return const vars that would be the way to go. I guess we could copy the state but right now the only place the state gets modified is during creation.

Reply:
Maybe as a follow on we could use copystructure or something to make a deep copy? For now this is probably fine.